src/sbg_device.cpp

```diff
@@ -0,0 +1,448 @@
+#include <iomanip>
+#include <fstream>
+#include <ctime>
+
+#include <boost/lexical_cast.hpp>
+#include <boost/regex.hpp>
+#include <boost/thread/xtime.hpp>
+#include <boost/date_time/local_time/local_time.hpp>
+
+#include "sbg_device.h"
+
+using namespace std;
+using sbg::SbgDevice;
+
+// From ros_com/recorder
+std::string timeToStr(ros::WallTime ros_t)
+{
+    (void)ros_t;
+    std::stringstream msg;
+    const boost::posix_time::ptime now = boost::posix_time::second_clock::local_time();
+    boost::posix_time::time_facet *const f = new boost::posix_time::time_facet("%Y-%m-%d-%H-%M-%S");
+    msg.imbue(std::locale(msg.getloc(),f));
+    msg << now;
+    return msg.str();
+}
+
+/*!
+ * Class to handle a connected SBG device.
+ */
+//---------------------------------------------------------------------//
+//- Constructor                                                       -//
+//---------------------------------------------------------------------//
+
+SbgDevice::SbgDevice(ros::NodeHandle* p_node_handle):
+m_p_node_(p_node_handle),
+m_mag_calibration_ongoing_(false),
+m_mag_calibration_done_(false)
+{
+  loadParameters();
+  connect();
+}
+
+SbgDevice::~SbgDevice(void)
+{
+  SbgErrorCode error_code;
+
+  error_code = sbgEComClose(&m_com_handle_);
+
+  if (error_code != SBG_NO_ERROR)
+  {
+    ROS_ERROR("Unable to close the SBG communication handle - %s.", sbgErrorCodeToString(error_code));
+  }
+
+  m_config_store_.closeCommunicationInterface(&m_sbg_interface_);
+}
+
+//---------------------------------------------------------------------//
+//- Private  methods                                                  -//
+//---------------------------------------------------------------------//
+
+SbgErrorCode SbgDevice::onLogReceivedCallback(SbgEComHandle* pHandle, SbgEComClass msgClass, SbgEComMsgId msg, const SbgBinaryLogData* pLogData, void* pUserArg)
+{
+  SbgDevice *p_sbg_device;
+  p_sbg_device = (SbgDevice*)(pUserArg);
+
+  p_sbg_device->onLogReceived(msgClass, msg, pLogData);
+
+  return SBG_NO_ERROR;
+}
+
+void SbgDevice::onLogReceived(SbgEComClass msgClass, SbgEComMsgId msg, const SbgBinaryLogData* pLogData)
+{
+  //
+  // Publish the received SBG log.
+  //
+  m_message_publisher_.publish(m_ros_processing_time_, msgClass, msg, *pLogData);
+}
+
+void SbgDevice::loadParameters(void)
+{
+  //
+  // Get the ROS private nodeHandle, where the parameters are loaded from the launch file.
+  //
+  ros::NodeHandle n_private("~");
+  m_config_store_.loadFromRosNodeHandle(n_private);
+}
+
+void SbgDevice::connect(void)
+{
+  SbgErrorCode error_code;
+
+  m_config_store_.initCommunicationInterface(&m_sbg_interface_);
```

Comment:
So in fact the store owns and is responsible of the interface / sbgECom protocol init
I would have preferred a better organized solution.
both sbgECom and sbgInterface should be handled / owned by the same C++ object and probably SbgDevice.
The store should just store the settings and then we should have an object to write/read a store to/from the device.